src/Workspaces/Core/Portable/Workspace/Solution/MetadataOnlyImage.cs

```diff
@@ -97,18 +102,15 @@ public MetadataReference CreateReference(ImmutableArray<string> aliases, bool em
                 // give them pointer to the native memory. also we need to handle lifetime ourselves.
                 var metadata = AssemblyMetadata.Create(ModuleMetadata.CreateFromImage(supportNativeMemory.GetPointer(), (int)stream.Length));
 
-                var referenceWithNativeMemory = metadata.GetReference(
+                // Tie lifetime of stream to metadata we created. It is important to tie this to the Metadata and not the
+                // metadata reference, as PE symbols hold onto just the Metadata.
+                s_lifetime.Add(metadata, supportNativeMemory);
```

Comment:
is there a race condition concern here?  .Add will throw if two people try to do this.  And isn't this code all free threaded?  shouild we be locking anywhere here?